mlir_graphblas/algorithms.py

```diff
@@ -2003,3 +2003,375 @@ def __call__(
+            row_index_i64 = irb.arith.index_cast(row_index, "i64")
+            n_times_row_index = irb.arith.muli(N, row_index)
+            n_times_row_index_i64 = irb.arith.index_cast(n_times_row_index, "i64")
+            irb.memref.store(n_times_row_index_i64, z_pointers, row_index)
+            row_vector_ptr = irb.llvm.getelementptr(z_vectors, row_index_i64)
+            row_vector_ptr8 = irb.llvm.load(row_vector_ptr, "!llvm.ptr<i8>")
+            row_vector = irb.util.ptr8_to_tensor(
+                row_vector_ptr8, "tensor<?xf64, #CV64>"
+            )
+            row_vector_values = irb.sparse_tensor.values(row_vector)
+            with irb.for_loop(0, N) as z_concat_loop_col_for_vars:
+                col_index = z_concat_loop_col_for_vars.iter_var_index
+                col_index_i64 = irb.arith.index_cast(col_index, "i64")
+                values_position = irb.arith.addi(n_times_row_index, col_index)
+                irb.memref.store(col_index_i64, z_indices, values_position)
+                value = irb.memref.load(row_vector_values, col_index)
+                irb.memref.store(value, z_values, values_position)
+        irb.memref.store(N_squared_i64, z_pointers, N)
+        # TODO ideally done with loop fusion with the above loops
+        z = irb.graphblas.apply(z, "cos")
+
+        c_vectors = irb.llvm.alloca(num_taus_i64, "!llvm.ptr<i8>")
+        two_over_N = irb.arith.divf(c2_f64, N_f64)
+        negative_a_over_lmax = irb.arith.divf(a, lmax)
+        negative_a_over_lmax = irb.arith.mulf(negative_a_over_lmax, c_negative_1_f64)
+        with irb.for_loop(0, num_taus) as tau_for_vars:
+            idx = tau_for_vars.iter_var_index
+            idx_i64 = irb.arith.index_cast(idx, "i64")
+            tau_ptr = irb.llvm.getelementptr(taus, idx_i64)
+            tau = irb.llvm.load(tau_ptr, "f64")
+            scalar_factor = irb.arith.mulf(tau, negative_a_over_lmax)
+            y = irb.graphblas.apply(num, "plus", right=c1_f64)
+            y = irb.graphblas.apply(y, "times", right=scalar_factor)
+            # TODO make "exp" an apply and update operator.
+            y = irb.graphblas.apply(y, "pow", left=c_e_f64)
+            y = irb.graphblas.apply(y, "times", right=two_over_N)
+            c_vector = irb.graphblas.matrix_multiply(y, z, "plus_times")
+            c_vector_ptr8 = irb.util.tensor_to_ptr8(c_vector)
+            destination_ptr = irb.llvm.getelementptr(c_vectors, idx_i64)
+            irb.llvm.store(c_vector_ptr8, destination_ptr)
+
+        n_features_out = num_taus
+        n_features_out_i64 = num_taus_i64
+        heat_print_matrices = irb.llvm.alloca(n_features_out_i64, "!llvm.ptr<i8>")
+
+        twf_old = signal
+        twf_cur = irb.graphblas.matrix_multiply(L, signal, "plus_times")
+        twf_cur = irb.graphblas.apply(twf_cur, "div", right=a)
+        # TODO this should be handled by irb.graphblas.update
+        twf_cur = irb.graphblas.union(twf_cur, signal, "minus")
```

Comment:
This is really a question for @eriknw : Why are we doing this? Are we sure it's safe to use the union?

I was hoping to avoid "minus" as an allowable operation for `union`, but this now requires that change. And we don't have the "require_monoid" logic forcing a user to opt-in.

If this is truly the correct code, I think I want to change this to `apply` with "ainv" on signal, and then a "plus" `union`.

Reply:
I'm not exactly sure of the correctness of the reference notebook implementation, but I remember verifying from the reference implementation that we wanted `[10, _, 20, _] - [2, 2, _, _]` to be `[8, 2, 20, _]` and not `[8, -2, 20, _]`. There's a chance I made a mistake in verifying, but this is what I recall.  I think using `ainv` with apply would lead to the undesired latter result.

Reply:
https://github.com/metagraph-dev/mlir-graphblas/commit/ce26f591efd6031e53380d12bb5641ce99a4d9bd implements this suggestion.